### ISSUE TITLE
docker healthcheck creates zombie processes if TLS is enabled with self signed certificate

### ISSUE THREAD

**Issue by hendrik1120** [authelia/authelia#7463]
### Version

v4.38.9

### Deployment Method

Docker

### Reverse Proxy

Traefik

### Reverse Proxy Version

3.0.3

### Description

Hi,

I would like to follow up on my Discord message and properly document this issue as a bug.

Although this issue is not within the Authelia application itself, it is within the Authelia codebase and is also present by default in every Docker image.

If Authelia has been configured with the TLS option enabled (see below), the Docker healthcheck command using `wget` will create a zombie process at every execution. The default Docker healthcheck interval is set to 30 seconds, which will create 2,880 processes per day (24 hours * 60 minutes/hour * 2 processes/minute). When using Debian as the Docker host, the process limit is around 7,000 processes per container. Therefore, the Authelia container will become unhealthy after about 2.5 days, leading Traefik to drop requests to this container.

In the rare case where the host OS (Unraid) does not limit the processes, the host will eventually run out of PIDs and crash.

### Reproduction

This issue can be replicated using the authelia docker image with the TLS option enabled in the configuration.yml.
```yaml
server:
  tls:
    key: '/config/private.pem'
    certificate: '/config/public.crt'
```
The key and certificate can be created using the command from the [docs](https://www.authelia.com/reference/guides/generating-secure-values/#generating-an-rsa-self-signed-certificate).
```bash
authelia crypto certificate rsa generate --common-name example.com --directory /config
```
To speed up the issue, I suggest increasing the healtheck interval using docker-compose:
```yaml
    healthcheck:
      interval: 0.01s
```

To check the amount of processes created, I ran `docker stats` on the host or `ps` inside the container.

### Expectations

The container healthcheck should not fail after a certain period of time.

### Configuration (Authelia)

```yaml
---
theme: dark

storage:
  encryption_key: 'a_very_important_secret'
  local:
    path: '/config/db.sqlite3'

log:
  level: trace

server:
  tls:
    key: '/config/private.pem'
    certificate: '/config/public.crt'

session:
  secret: 'insecure_session_secret'
  cookies:
    - domain: 'example.com'
      authelia_url: 'https://auth.example.com'
      default_redirection_url: 'https://www.example.com'
      name: 'authelia_session'

notifier:
  disable_startup_check: false
  filesystem:
    filename: '/config/notification.txt'

access_control:
  default_policy: 'one_factor'

identity_validation:
  reset_password:
    jwt_lifespan: '5 minutes'
    jwt_algorithm: 'HS256'
    jwt_secret: 'insecure_secret'

authentication_backend:
  file:
    path: '/config/users.yml'
    watch: false
    search:
      email: false
      case_insensitive: false
    password:
      algorithm: 'argon2'
...
```

### Build Information

```shell
Last Tag: v4.38.9
State: tagged clean
Branch: v4.38.9
Commit: 2798576ee25a56fd4c14814acd087b6e92f3978b
Build Number: 30034
Build OS: linux
Build Arch: amd64
Build Compiler: gc
Build Date: Sun, 16 Jun 2024 19:47:16 +1000
Extra: 

Go:
    Version: go1.22.2
    Module Path: github.com/authelia/authelia/v4
    Executable Path: github.com/authelia/authelia/v4/cmd/authelia
```

### Logs (Authelia)

```shell
time="2024-06-28T15:39:34+02:00" level=debug msg="Loaded Configuration Sources" caller="github.com/authelia/authelia/v4/internal/commands/context.go:266 (*CmdCtx).LogConfigure" files="[/config/configuration.yml]" filters="[template]"
time="2024-06-28T15:39:34+02:00" level=debug msg="Logging Initialized" caller="github.com/authelia/authelia/v4/internal/commands/context.go:267 (*CmdCtx).LogConfigure" fields.level=trace file= format= keep_stdout=false
time="2024-06-28T15:39:34+02:00" level=debug msg="Process user information" caller="github.com/authelia/authelia/v4/internal/commands/context.go:307 (*CmdCtx).LogProcessCurrentUserRunE" gid=99 uid=99
time="2024-06-28T15:39:34+02:00" level=warning msg="Configuration: access_control: no rules have been specified so the 'default_policy' of 'one_factor' is going to be applied to all requests" caller="github.com/authelia/authelia/v4/internal/commands/context.go:317 (*CmdCtx).ConfigValidateLogRunE"
time="2024-06-28T15:39:34+02:00" level=info msg="Authelia v4.38.9 is starting" caller="github.com/authelia/authelia/v4/internal/commands/root.go:62 (*CmdCtx).RootRunE"
time="2024-06-28T15:39:34+02:00" level=info msg="Log severity set to trace" caller="github.com/authelia/authelia/v4/internal/logging/logger.go:93 setLevelStr"
time="2024-06-28T15:39:34+02:00" level=trace msg="Starting scan of directory  for certificates" caller="github.com/authelia/authelia/v4/internal/utils/crypto.go:350 NewX509CertPool"
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:105 doStartupChecks" provider=storage
time="2024-06-28T15:39:34+02:00" level=info msg="Storage schema is being checked for updates" caller="github.com/authelia/authelia/v4/internal/storage/sql_provider.go:338 (*SQLProvider).StartupCheck"
time="2024-06-28T15:39:34+02:00" level=info msg="Storage schema is already up to date" caller="github.com/authelia/authelia/v4/internal/storage/sql_provider.go:356 (*SQLProvider).StartupCheck"
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:112 doStartupChecks" provider=storage
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:115 doStartupChecks" provider=user
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:122 doStartupChecks" provider=user
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:125 doStartupChecks" provider=notification
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:132 doStartupChecks" provider=notification
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:135 doStartupChecks" provider=ntp
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:146 doStartupChecks" provider=ntp
time="2024-06-28T15:39:34+02:00" level=trace msg="Starting Services" caller="github.com/spf13/cobra@v1.8.1/command.go:985 (*Command).execute"
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:222 handleRouter" implementation=Legacy methods="*" path_prefix=/api/verify
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:242 handleRouter" implementation=AuthRequest methods="[GET HEAD]" path=/api/authz/auth-request
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:233 handleRouter" implementation=ExtAuthz methods="*" path_prefix=/api/authz/ext-authz
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:242 handleRouter" implementation=ForwardAuth methods="[GET HEAD]" path=/api/authz/forward-auth
time="2024-06-28T15:39:34+02:00" level=trace msg="Service Loaded" caller="github.com/authelia/authelia/v4/internal/commands/services.go:319 servicesRun" server=main service=server
time="2024-06-28T15:39:34+02:00" level=debug msg="Create Server Service (metrics) skipped" caller="github.com/authelia/authelia/v4/internal/commands/services.go:318 servicesRun"
time="2024-06-28T15:39:34+02:00" level=info msg="Startup complete" caller="github.com/authelia/authelia/v4/internal/commands/root.go:94 (*CmdCtx).RootRunE"
time="2024-06-28T15:39:34+02:00" level=info msg="Listening for TLS connections on '[::]:9091' path '/'" caller="github.com/authelia/authelia/v4/internal/commands/services.go:137 (*ServerService).Run" server=main service=server
root@debian:~/docker# docker logs authelia
time="2024-06-28T15:39:34+02:00" level=debug msg="Loaded Configuration Sources" caller="github.com/authelia/authelia/v4/internal/commands/context.go:266 (*CmdCtx).LogConfigure" files="[/config/configuration.yml]" filters="[template]"
time="2024-06-28T15:39:34+02:00" level=debug msg="Logging Initialized" caller="github.com/authelia/authelia/v4/internal/commands/context.go:267 (*CmdCtx).LogConfigure" fields.level=trace file= format= keep_stdout=false
time="2024-06-28T15:39:34+02:00" level=debug msg="Process user information" caller="github.com/authelia/authelia/v4/internal/commands/context.go:307 (*CmdCtx).LogProcessCurrentUserRunE" gid=99 uid=99
time="2024-06-28T15:39:34+02:00" level=warning msg="Configuration: access_control: no rules have been specified so the 'default_policy' of 'one_factor' is going to be applied to all requests" caller="github.com/authelia/authelia/v4/internal/commands/context.go:317 (*CmdCtx).ConfigValidateLogRunE"
time="2024-06-28T15:39:34+02:00" level=info msg="Authelia v4.38.9 is starting" caller="github.com/authelia/authelia/v4/internal/commands/root.go:62 (*CmdCtx).RootRunE"
time="2024-06-28T15:39:34+02:00" level=info msg="Log severity set to trace" caller="github.com/authelia/authelia/v4/internal/logging/logger.go:93 setLevelStr"
time="2024-06-28T15:39:34+02:00" level=trace msg="Starting scan of directory  for certificates" caller="github.com/authelia/authelia/v4/internal/utils/crypto.go:350 NewX509CertPool"
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:105 doStartupChecks" provider=storage
time="2024-06-28T15:39:34+02:00" level=info msg="Storage schema is being checked for updates" caller="github.com/authelia/authelia/v4/internal/storage/sql_provider.go:338 (*SQLProvider).StartupCheck"
time="2024-06-28T15:39:34+02:00" level=info msg="Storage schema is already up to date" caller="github.com/authelia/authelia/v4/internal/storage/sql_provider.go:356 (*SQLProvider).StartupCheck"
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:112 doStartupChecks" provider=storage
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:115 doStartupChecks" provider=user
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:122 doStartupChecks" provider=user
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:125 doStartupChecks" provider=notification
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:132 doStartupChecks" provider=notification
time="2024-06-28T15:39:34+02:00" level=trace msg="Performing Startup Check" caller="github.com/authelia/authelia/v4/internal/commands/root.go:135 doStartupChecks" provider=ntp
time="2024-06-28T15:39:34+02:00" level=trace msg="Startup Check Completed Successfully" caller="github.com/authelia/authelia/v4/internal/commands/root.go:146 doStartupChecks" provider=ntp
time="2024-06-28T15:39:34+02:00" level=trace msg="Starting Services" caller="github.com/spf13/cobra@v1.8.1/command.go:985 (*Command).execute"
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:222 handleRouter" implementation=Legacy methods="*" path_prefix=/api/verify
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:242 handleRouter" implementation=AuthRequest methods="[GET HEAD]" path=/api/authz/auth-request
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:233 handleRouter" implementation=ExtAuthz methods="*" path_prefix=/api/authz/ext-authz
time="2024-06-28T15:39:34+02:00" level=trace msg="Registering Authz Endpoint" caller="github.com/authelia/authelia/v4/internal/server/handlers.go:242 handleRouter" implementation=ForwardAuth methods="[GET HEAD]" path=/api/authz/forward-auth
time="2024-06-28T15:39:34+02:00" level=trace msg="Service Loaded" caller="github.com/authelia/authelia/v4/internal/commands/services.go:319 servicesRun" server=main service=server
time="2024-06-28T15:39:34+02:00" level=debug msg="Create Server Service (metrics) skipped" caller="github.com/authelia/authelia/v4/internal/commands/services.go:318 servicesRun"
time="2024-06-28T15:39:34+02:00" level=info msg="Startup complete" caller="github.com/authelia/authelia/v4/internal/commands/root.go:94 (*CmdCtx).RootRunE"
time="2024-06-28T15:39:34+02:00" level=info msg="Listening for TLS connections on '[::]:9091' path '/'" caller="github.com/authelia/authelia/v4/internal/commands/services.go:137 (*ServerService).Run" server=main service=server
time="2024-06-28T15:39:39+02:00" level=trace msg="Request hit" caller="github.com/authelia/authelia/v4/internal/middlewares/log_request.go:12 handleRouter.LogRequest.func40" method=GET path=/api/health remote_ip="::1"
time="2024-06-28T15:39:39+02:00" level=trace msg="Replied (status=200)" caller="github.com/authelia/authelia/v4/internal/middlewares/log_request.go:16 handleRouter.LogRequest.func40" method=GET path=/api/health remote_ip="::1"
```

### Documentation

https://www.authelia.com/configuration/miscellaneous/server/#tls
https://www.authelia.com/overview/security/measures/#mutual-tls

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)

- [X] This is a bug report and not a support request

- [X] I have read the security policy and this bug report is not a security issue or security related issue

- [X] I have either included the complete configuration file or I am sure it's unrelated to the configuration

- [X] I have either included the complete debug / trace logs or the output of the build-info command if the logs are not relevant

- [X] I have provided all of the required information in full with the only alteration being reasonable sanitization in accordance with the [Troubleshooting Sanitization](https://www.authelia.com/r/sanitize) reference guide

- [X] I have checked for related proxy or application logs and included them if available

- [X] I have checked for related issues and checked the documentation

**Comment by james-d-elliott** [comment]
Looks to be a bug with busybox wget's ssl_client. This should fix it https://github.com/authelia/authelia/pull/7498

**Comment by hendrik1120** [comment]
Can confirm, #7498 does fix the issue.

LimeTech (creator of UNRAID) is also working on a fix for their docker implementation to prevent process leaks from crashing the host.

**Comment by james-d-elliott** [comment]
Great to know. Can't really see why the busybox wget has not been fixed. The issue occurs with any TLS request it looks like.

**Comment by hendrik1120** [comment]
Yeah, just found this:  https://bugs.busybox.net/show_bug.cgi?id=15967

Adding `--init`to the docker run command has been suggested, but I really don't think that is an acceptable solution.

**Comment by james-d-elliott** [comment]
Hmm, I'm kind of inclined to say it may be the correct solution and I may revert it, and instead update the docs. Considering the only instance where this has been discovered is an edge case it'd make more sense to do that then to force everyone to download a larger image.

Init runs tini, which is included with docker older than 1.24. I will do some more research I think.

**Comment by hendrik1120** [comment]
Yeah, the whole thing kinda is an edge case on an edge case and is very unlikely to cause these massive problems to anyone else.
I am just using TLS between authelia and traefik. It's probably overkill, but “For the best security protection, configuration with TLS is highly recommended. …” from the [docs](https://www.authelia.com/overview/security/measures/#mutual-tls).

I agree with you, but if someone forgets to set this uncommon docker setting, it could be really hard to troubleshoot. I am not sure how predictable the authelia app itself fails when being PID limited, which might cause more edge cases (even security related maybe?). I did manage to break authelia very badly in one case. It also depends on the os and if the docker compose `pids_limit` is set.

**Comment by james-d-elliott** [comment]
I'll update the docs on that to clarify secure networks vs insecure ones. i.e. a iptables network on a host is significantly more secure as they'd likely have to compromise a host with root equivalent access as well to compromise that network, making the effort inconsequential in a vast majority of cases since they'd be able to alter configs themselves if they have root equivalent access.

**Comment by hendrik1120** [comment]
Resolved in v4.38.10
